### DISCUSSION TITLE
State Management Challenges

### DISCUSSION THREAD

**Discussion by upromeer** [pmndrs/zustand#2391] [Q&A]
Hey! I'm currently facing some challenges with state management in Zustand, and I could use some guidance to clear up my doubts and resolve errors. Here's a breakdown of my setup:

using Next.js with App Router, TypeScript, React-Query for global state management, Zustand for local state management with persist, immer, and devtool integration

- Re-rendering Issue in Components:
I'm encountering re-rendering issues in various components, specifically within `src/app/list/Comp`. In this scenario, I've adopted a selective approach, where **Side2** contains a controlled input. Interestingly, handling the input in **Side2** triggers re-renders in **Side1**, **ControlBtnPrev**, and **ControlBtnNext**. Notably, I'm avoiding the use of a **saved** state. Any insights on what might be causing this behavior?

- Undefined Error with setLists on /list Route:
Moving to `src/app/list/page`, I'm facing issues with the setLists function when hitting the `/list` route. If any component related to state is used, it throws an error mentioning undefined. How can I address this issue with **setLists**?

- Optimal Usage of React-Query:
I'm currently utilizing React-Query inside `src/app/list/page` by running effects. However, I'm curious about the best practice—should I continue using it within components, or would it be more appropriate to integrate it into the **StoreProvider** and pass props to **createStore**?

I suspect there might be additional glitches in my implementation. To provide more context and get specific assistance, here's the link to the code: https://codesandbox.io/p/devbox/rand-dev-box-xypvx6. _Stuff are inside `/app/list` and `/stores`_
I would greatly appreciate any guidance on what I might be doing wrong or how I can improve my setup.

**Comment by upromeer** [comment]
please, have a look ...

**Reply by dbritto-dev** [reply]
@upromeer let me help you to tackle some issues in your poc:

- Re-rendering issues: those issues are related to use selectors without equality function so instead of `useStore` you should use `useStoreWithEqualityFn`
- React Query: the usage of `QueryClientProvider` is a must if you want to use any hook like `useQuery` or `useMutation`. Please read this guide -> https://tanstack.com/query/v4/docs/framework/react/guides/ssr
- List page crash: that's expected since lists is empty so you need to be careful due to list being undefined

**Reply by upromeer** [reply]
u cant get my point 2,3.
can u please, have a look in provided codesandbox code and fix those issue.

**Reply by dbritto-dev** [reply]
@upromeer please, review my answer and apply the suggestions

**Reply by upromeer** [reply]
list pages crash due to list is empty.
but i want to fetch data using react-query and pass to store as initialProp in createStore(init(data))
so that it wont crash
or i making state as empty and on page streaming.. with react-query and setData(data)

i dont have to fetch anything on ssr.

**Reply by dbritto-dev** [reply]
@upromeer even if you can pass the initial data to your store since is a list you should always consider that lists can be an empty array

**Reply by upromeer** [reply]
no it shouldnt, have always data inside it.
user just navigate and updated saved list (not fetched list)

user load page
data - fetched by react-query or fetch (must be cached)
there is suspense fallback
as data present
actual content renders
using state in multiple component, (must prevent extra re-renders) while updating state

**Reply by dbritto-dev** [reply]
well that doesn't sound safe to me, but up to you

**Reply by upromeer** [reply]
suggest me then, what are ways i should follow.
end goal is making my app performent as possible in both end (user x + dev x)

**Reply by dbritto-dev** [reply]
I thought I answered all your questions the rest is up to you

**Reply by upromeer** [reply]
i only get re-rendering solution :cry:

**Reply by dbritto-dev** [reply]
- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-fetch
- https://tanstack.com/query/v4/docs/framework/react/guides/ssr#using-initialdata-2
- https://docs.pmnd.rs/zustand/guides/nextjs

**Comment by vaynevayne** [comment]
Are there real status management challenges for websites?
